Ensure no ALPN is negotiated for SMTP

mvdnes opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Ensure no ALPN is sent for SMTP connections. Should fix #138573.

The httpcore library used by the home_connect component also calls `set_alpn_protocols` with a value of `['http/1.1']`. This interfers with the SMTP notify feature, causing it to fail in setting up a connection.

I have no idea why the contexts interfere with eachother, and believe proper isolation would be a better fix. On the other hand, this fix is easy enough and should fix notifications again for some people.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #138573
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mvdnes committed:
Ensure no ALPN is negotiated for SMTP

bdraco commented on homeassistant/components/smtp/notify.py:
```diff
@@ -141,6 +141,7 @@ def connect(self):
         """Connect/authenticate to SMTP Server."""
         ssl_context = client_context() if self._verify_ssl else None
         if self.encryption == "tls":
+            ssl_context.set_alpn_protocols([])
```
This looks like its changing the context for all instances?

mvdnes replied:
Well yes, but this line actually resets it to the default value.

bdraco replied:
Please make a new context in the executor or save it in a named constant at import time.

mvdnes replied:
Thank you for the feedback. As httpx_client also changes the alpn_protocols, I am unsure what the best course of action is. I will first try to discuss this in https://github.com/home-assistant/core/issues/138573#issuecomment-2781281784.

mvdnes replied:
I have created a new context as per your suggestion.

mvdnes committed:
Ensure smtp.notify has an independent SSLContext

Normally, ssl.client_context is shared between components. However, some
other components may call set_alpn_protocols, which may make the SMTP
connection fail. By having an independent SSLContext, we ensure that
this component is not affected by other code.

mvdnes commented:
I have changed my approach to creating an independent SSLContext for this component.

bdraco commented on homeassistant/util/ssl.py:
```diff
@@ -139,6 +147,14 @@ def client_context(
     return _SSL_CONTEXTS.get(ssl_cipher_list, _DEFAULT_SSL_CONTEXT)
 
 
+def create_client_context(
```
Please do this change in a separate PR. We shouldn't change core utils in the same per as an integration change

mvdnes replied:
Ok, I opened #142653 and will rebase this once it is merged.

bdraco commented on homeassistant/components/smtp/notify.py:
```diff
@@ -136,23 +136,23 @@ def __init__(
         self.debug = debug
         self._verify_ssl = verify_ssl
         self.tries = 2
+        self._ssl_context = create_client_context() if self._verify_ssl else None
```
Please avoid side effects in `__init__`

Instead create this in `get_service` and pass it in

mvdnes committed:
Move ssl_context initializer to get_service

mvdnes committed:
Merge branch 'dev' into smtp-notify-alpn

mvdnes commented:
The code for util.ssl is now in dev, so now all changes are only to the SMTP component.

bdraco commented:
It would be nice to cover the missing line merging, but OK if you can't since its only changing out the context

mvdnes commented:
I see that all of the `connect` and `connection_is_valid` code is not covered by tests.
Since that would require setting up a server, or mocking the server connection, I think it would be best to merge this as-is. I do not have a lot of experience in setting up those kinds of tests.

bdraco commented:
Thanks @mvdnes

mvdnes commented:
@bdraco Thank you too, for the reviews and guidance.
